### DISCUSSION TITLE
Conflicting tokens not resolved by token match length

### DISCUSSION THREAD

**Discussion by Liby99** [tree-sitter/tree-sitter#1357] [Q&A] [answered]
Hello! I've been working on a grammar and there is an ambiguity that I think should be disambiguated by lexical analysis (token length) but it is actually not. Please allow me to introduce the (simplified) language. The language is just recognizing simple expressions. Here are some examples of expressions:

```
// identifier
a

// 3-tuple
a, b, c

// qualified identifier (a token)
module:member
module:submodule:member

// abstraction which reads "for parameter `a` and `b`, id `c`"
a, b: c

// abstraction which reads "for parameter `a`, tuple `(b, c)`"
a: b, c
```

Now, the conflict comes up when we have

```
a, b:c
```

A correct interpretation for this expression is treating it as a tuple, where the first element is an id `"a"`, and the second element is a qualified id `"b:c"`

```wat
// Tree A
(top_level [0, 0] - [1, 0]
  (expr [0, 0] - [0, 6]
    (tuple_expr [0, 0] - [0, 6]
      (tuple_expr [0, 0] - [0, 1]
        (basic_expr [0, 0] - [0, 1]
          (id [0, 0] - [0, 1])))
      (basic_expr [0, 3] - [0, 6]
        (qualified_id [0, 3] - [0, 6])))))
```

However, another (incorrect) parse is possible, where it is recognized as an abstraction -- `"a"` and `"b"` are recognized as the formal parameters and `"c"` is the expression to evaluate in the abstraction.

```wat
// Tree B
(top_level [0, 0] - [1, 0]
  (expr [0, 0] - [0, 6]
    (abstract_expr [0, 0] - [0, 6]
      (bindings [0, 0] - [0, 4]
        (formal_param [0, 0] - [0, 1])
        (formal_param [0, 3] - [0, 4]))
      (basic_expr [0, 5] - [0, 6]
        (id [0, 5] - [0, 6])))))
```

I think this could be resolved by token length, because the token `b:c` is longer than the token `b`. Therefore we should be selecting `Tree A` over `Tree B`. But unfortunately it is producing two conflicting trees and have to select one arbitrarily. I am aware of dynamic precedence levels but I don't really want to use that feature (due to other parts of the language).

I wonder what could be a solution to this problem?

---

The full tree-sitter grammar is listed below:

```js
identifier = /[a-zA-Z][a-zA-Z]*/

module.exports = grammar({
    name: 'MYGRAMMAR',
    extras: $ => [/\s/],
    conflicts: $ => [
        [$.id, $.formal_param],
    ],
    rules: {
        top_level: $ => $.expr,

        // Expressions
        expr: $ => choice(
            $.tuple_expr,
            $.abstract_expr,
        ),
        abstract_expr: $ => seq($.bindings, ':', $.basic_expr),
        tuple_expr: $ => choice(
            $.basic_expr,
            seq($.tuple_expr, ',', $.basic_expr),
        ),
        basic_expr: $ => choice(
            $.id,
            $.qualified_id,
        ),

        // Bindings
        formal_param: $ => token(identifier),
        bindings: $ => seq($.formal_param, repeat(seq(',', $.formal_param))),

        // Identifiers
        qualified_id: $ => token(/[a-zA-Z]+:[a-zA-Z]+(:[a-zA-Z]+)/),
        id: $ => token(identifier),
    },
});
```

**Comment by ahelwer** [comment]
Consider the conflict `[$.tuple_expr, $.abstract_expr]`; it occurs because both of those rules have the `$.basic_expr` rule as a choice. I don't know what you want the behavior to be, but by declaring that a conflict you are telling tree-sitter to resolve it using GLR parsing. Thus tree-sitter uses GLR parsing to find both possible parses of `a, b:c` (you can see this happen using the `tree-sitter parse -d` flag) and chooses one arbitrarily (in this case the one you don't want). Given a `(basic_expr)` token, should tree-sitter parse it as `(tuple_expr (basic_expr))` or `(abstract_expr (basic_expr))`? You need to find the answer to that question, and rewrite your grammar in such a way that you don't need to declare the conflict.

**Reply by Liby99** [reply]
Ah thanks for pointing out. Sorry that is actually a mistake made by me when I was simplifying the grammar. The `basic_expr` should not be a part of `abstract_expr` actually. So let me update the tree-sitter grammar file in the original post.

**Reply by ahelwer** [reply]
@Liby99 You've slain one of the conflicts, now you must also get rid of the other!

**Reply by ahelwer** [reply]
@Liby99 is it correct to say that if you see the sequence `a:b` it should *never* be parsed as an `abstract_expr`? So `abstract_expr` is only valid if there are spaces, like `a :b` or something like that? If no such differentiator exists, this is an authentic ambiguity in the language and must be resolved with dynamic precedence.

**Reply by Liby99** [reply]
Yeah if we see a sequence `a:b` it should be recognized as a single token of `qualified_id` and therefore will never be parsed as an `abstract_expr`.

**Reply by mbravenboer** [reply]
@ahelwer yeah, that is correct, but take for example `3.14`, which can also be parsed as a field access or a floating-point literal. Afaik, this is lexically disambiguated in tree-sitter. That's also an authentic ambiguity if that lexical disambiguation does not happen. I'm not sure why that does not kick in here.

**Reply by Liby99** [reply]
If there is space as in `a, b: c` or `a :b` then, yes, they should be recognized as `abstract_expr`.

**Reply by ahelwer** [reply]
@Liby99 what are the actual rules around spaces? You'll have to add explicit spaces to your definition of `abstract_expr`. Like `abstract_expr: $ => seq($.bindings, choice(' :', ': '), $.basic_expr)`

**Reply by ahelwer** [reply]
@mbravenboer the lexical rule has nothing to do with it. Both parses of `a, b:c` are valid. You need to choose between them at the parser level, not the lexer level. Your float example works when you're looking a single token, but we're looking at a sequence of tokens here. edit: the lexical rule *does* have something to do with it! It's just that context-aware lexing takes precedence over match length.

**Reply by Liby99** [reply]
Isn't `qualified_id: $ => token(seq(identifier, ':', identifier, repeat(seq(':', identifier))))` declaring that `qualified_id` should be a single token?

I have also tried using this

```
qualified_id: $ => token(/[a-zA-Z]+:[a-zA-Z]+(:[a-zA-Z]+)/)
```

but it is **NOT** solving the problem. Actually let me update the original post so that we don't confuse about `qualified_id` being a sequence of tokens.

**Reply by ahelwer** [reply]
@mbravenboer to expand - this is a somewhat unusual (by the standards of other parsers) behavior of tree-sitter here, which is that if it expects a token and then it finds a prefix of the next lexeme satisfying that token, it will split up the lexeme to try to find a good parse. You can see this behavior described [here](https://tree-sitter.github.io/tree-sitter/creating-parsers#keywords) in a somewhat analogous situation.

**Reply by ahelwer** [reply]
@Liby99 I tried using `abstract_expr: $ => seq($.bindings, choice(' :', ': '), $.basic_expr)` and it produces the behavior you want. You are making the assumption that specifying `qualified_id` as a single token prevents the string `a:b` from *ever* being interpreted as anything other than a `qualified_id`, but that is not true. Tree-sitter uses context-aware lexing which means that it can split up `a:b` into `a`, `:`, and `b` if that would produce a valid parse for the GLR branch that it's in, which is what it does.

**Reply by Liby99** [reply]
Maybe this is complicating things but I do have expressions like `1 + 2` in the complete language. So the expression `a:1+a` should be totally valid here, recognized as `abstract_expr`. If we are posing strict whitespace around `":"` I think it is a bit overkill.

**Reply by ahelwer** [reply]
@Liby99 that does complicate things quite a bit. What language is this? It would help to find the formal semantics governing how these expressions are lexed and parsed.

**Reply by mbravenboer** [reply]
@ahelwer I appreciate the help (and I'm familiar with scannerless parsing and context-aware lexical analysis), but I don't see yet how what you're explaining clarifies why 3.14 is not ambiguous in a basic tree-sitter grammar. Am I missing something there? It seems like the exact same situation to me.

We were expecting this to apply:

> Match Length - If multiple valid tokens with the same precedence match the characters at a given position in a document, Tree-sitter will select the token that matches the longest sequence of characters.

(and generally we've exhausted experiments with the options in https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens )

**Reply by mbravenboer** [reply]
Is it maybe that the match length conflict resolution only works if you're not in two different GLR branches? That would explain it, but is also a very procedural way of handling that match length rule?

**Reply by ahelwer** [reply]
@mbravenboer the tricky thing here is realizing that by the time tree-sitter reaches the `b` in `a, b:c` it is in one of two GLR parse branches. In the first, the one you want, tree-sitter is assuming it's inside a `(tuple_expr (basic_expr))` node so the `b:c` will be lexed as a single `qualified_id` token because `qualified_id` is a valid token at that position and it has a longer match length than an `id` token. In the second branch, tree-sitter is assuming it's inside an `(abstract_expr (bindings))` node so it is looking for a `formal_param` (`qualified_id` is not valid here). Following the context-aware lexing rule (which has [higher precedence](https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens) than the match length rule) it tokenizes `b` as a `formal_param`. So I was wrong up above when I said it had nothing to do with the lexical rules! I forgot context-aware lexing was an actual rule in the list. Anyway, since both GLR parse branches produce valid parses, you need to tell tree-sitter how to choose between them. One way of doing this is to make the `abstract_expr` branch not parse correctly by explicitly adding spaces. Another is to use dynamic precedence.

For the 3.14 example, usually floats are terminals in an expression rule (as might be `a.b` field values). In such a case *both* floats and field values would be valid tokens, so context-aware lexing wouldn't be able to choose one over the other. It would fall back to one of the other lexing rules, perhaps longest match.

**Reply by mbravenboer** [reply]
@ahelwer thanks, yeah, I think that is exactly clear and must be the explanation. Thanks!

In a previous project I used a different GLR parser (SGLR - Scannerless GLR) and it has a notion of 'follow restriction' for lexical disambiguation. A follow restriction for example says that an identifier token cannot be followed by an identifier character. This restriction is not only used for conflict resolution (as in tree-sitter), but applies to whatever state the parser may be in (and also rejects parses if no other valid parse exists).

My confusion (I work with @Liby99 ) comes from the assuming that the tree-sitter lexical analysis works similarly, but from imaging how it is implemented I now understand how the conflict resolution applies.

SGLR also has 'reject' productions to declaratively state that a keyword cannot be used as an identifier. Both these mechanisms could maybe be useful in tree-sitter.

**Reply by ahelwer** [reply]
@mbravenboer ah yeah tree-sitter doesn't have any built-in way to specify such a follow restriction. If at the end of the day you really require complete control over lexing of identifiers (or other things), the heavyweight solution is to move their lexing logic into an external scanner where you have arbitrary lookahead (and lookbehind, if you store state). However, I went [a fair ways down that road](https://github.com/tree-sitter/tree-sitter/discussions/1257) and it gets real complicated real quick.

**Comment by ahlinc** [comment, marked as answer]
@Liby99 Something like bellow seems working for your described cases:
```js
identifier = /[a-zA-Z][a-zA-Z]*/

module.exports = grammar({
    name: 'MYGRAMMAR',
    extras: $ => [/\s/, $.comment],
    conflicts: $ => [
        [$.id, $.formal_param],
        [$.abstract_expr, $._tuple_item],
    ],
    rules: {
        top_level: $ => repeat($.expr),

        // Comments
        comment: $ => /\/\/.*/,

        // Expressions
        expr: $ => choice(
            prec.dynamic(1, $.tuple_expr),
            $.abstract_expr,
        ),
        abstract_expr: $ => seq(
            $.bindings, ':',
            choice(
                $.basic_expr,
                $.tuple_expr,
            ),
        ),
        tuple_expr: $ => $._tuple_item,
        _tuple_item: $ => choice(
            $.basic_expr,
            seq($._tuple_item, ',', $.basic_expr),
        ),
        basic_expr: $ => choice(
            $.id,
            $.qualified_id,
        ),

        // Identifiers
        qualified_id: $ => /[a-zA-Z]+:[a-zA-Z]+(:[a-zA-Z]+)?/,
        id: $ => identifier,

        // Bindings
        formal_param: $ => identifier,
        bindings: $ => seq($.formal_param, repeat(seq(',', $.formal_param))),
    },
});
```

<details><summary><i>Screenshot:</i></summary>

![Screenshot from 2021-08-25 11-41-28](https://user-images.githubusercontent.com/14666676/130758410-17b41ecc-2f5b-4beb-8cd8-18568c114278.png)

</details>
